src/lib/include/vector_triple.h

```diff
@@ -0,0 +1,135 @@
+#ifndef DISTOPIA_VECTOR_TRIPLE
+#define DISTOPIA_VECTOR_TRIPLE
+
+#include "compiler_hints.h"
+#include "distopia_type_traits.h"
+#include "x86_swizzle.h"
+#include "x86_tgintrin.h"
+#include "x86_vectors.h"
+
+// VectorTriple base class packs 3xSIMD datatypes into a single class.
+// Can be constructed from 3 x VectorT.
+// Can also be constructed from a ScalarT array which loads the right number
+// of values into the 3 x VectorT.
+template <typename VectorT> class VectorTriple {
+public:
+  // maps to vectorT to ScalarT
+  using ScalarT = VectorToScalarT<VectorT>;
+  // 3x SIMD Datatypes
+  VectorT x;
+  VectorT y;
+  VectorT z;
+  // number of values in the packed into the whole 3 x VectorT struct.
+  constexpr static std::size_t n_scalars = ValuesPerPack<VectorT> * 3;
+
+  // default construct
+  VectorTriple() = default;
+
+  // construct from 3 SIMD Vector datatypes eg __m128 or __m128d
+  inline VectorTriple(const VectorT a, const VectorT b, const VectorT c)
+      : x(a), y(b), z(c) {}
+
+  // construct by loading from an array of ScalarT eg float* or double *.
+  inline VectorTriple(const ScalarT *source)
+      : x(loadu_p<VectorT>(source)),
+        y(loadu_p<VectorT>(&source[ValuesPerPack<VectorT>])),
+        z(loadu_p<VectorT>(&source[+2 * ValuesPerPack<VectorT>])) {}
+
+  // construct by loading discontiguously from an array of ScalarT eg float* or
+  // double* for which the SIMD width is 4 (__m128 and __m256d). The access is
+  // indexed by 4 integers i,j,k,l where each index is the number
+  // of the particle. Assumes the input vector is in AOS format and returns SOA
+  inline VectorTriple(ScalarT *source, ScalarT *end, int i, int j, int k,
+                      int l) {
+    static_assert(ValuesPerPack<VectorT> == 4,
+                  "Cannot use this constructor on a type "
+                  "that does not have a SIMD width of 4");
+    VectorT a = SafeIdxLoad4<VectorT>(source, 3 * i, end);
+    VectorT b = SafeIdxLoad4<VectorT>(source, 3 * j, end);
+    VectorT c = SafeIdxLoad4<VectorT>(source, 3 * k, end);
+    VectorT d = SafeIdxLoad4<VectorT>(source, 3 * l, end);
+    // deinterleave
+    Deinterleave4x3(a, b, c, d, this->x, this->y, this->z);
+  }
+
+  // construct by loading discontiguously from an array of ScalarT eg float* or
+  // double* for which the SIMD width is 8 (__m256). The access is
+  // indexed by 8 integers i,j,k,l,m,n,o,p where each index is the number
+  // of the particle. Assumes the input vector is in AOS format and returns SOA
+  inline VectorTriple(ScalarT *source, ScalarT *end, int i, int j, int k, int l,
```

Comment:
I think we can kill this function and just use the size_t* version

Reply:
By this you mean pass in an array of indicies and read off that right? We then increment the pointer to the index array?